pkg/pod/entrypoint_lookup_impl_test.go

```diff
@@ -0,0 +1,172 @@
+	// Request authentication for ping request
+	// For further reference see https://docs.docker.com/registry/spec/api/#api-version-check
+	if r.URL.Path == "/v2/" && r.Method == http.MethodGet {
+		w.Header().Add("WWW-Authenticate", "basic")
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+	// Check auth if we've fetching the image
+	if strings.HasPrefix(r.URL.Path, "/v2/task") && r.Method == "GET" {
+		u, p, ok := r.BasicAuth()
+		if !ok || username != u || password != p {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+	}
+	// Default to open
+	f.reg.ServeHTTP(w, r)
+}
+
+func generateSecret(host string, username string, password string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "secret",
+			Namespace: "ns",
+		},
+		Type: corev1.SecretTypeDockercfg,
+		Data: map[string][]byte{
+			corev1.DockerConfigKey: []byte(
+				fmt.Sprintf(`{%q: {"auth": %q}}`,
+					host,
+					base64.StdEncoding.EncodeToString([]byte(username+":"+password)),
+				),
+			),
+		},
+	}
+}
+
+func TestGetImageWithImagePullSecrets(t *testing.T) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	ftp := newfakeHTTP()
+	s := httptest.NewServer(&ftp)
+	defer s.Close()
+
+	u, err := url.Parse(s.URL)
+	if err != nil {
+		t.Errorf("Parsing url with an error: %v", err)
+	}
+
+	task := parse.MustParseTask(t, `metadata:name: test-create-image`)
+
+	ref, err := remotetest.CreateImageWithAnnotations(u.Host+"/task/test-create-image", remotetest.DefaultObjectAnnotationMapper, task)
+	if err != nil {
+		t.Errorf("uploading image failed unexpectedly with an error: %v", err)
+	}
+
+	imgRef, err := name.ParseReference(ref)
+	if err != nil {
+		t.Errorf("digest %s is not a valid reference: %v", ref, err)
+	}
+
+	for _, tc := range []struct {
+		name             string
+		basicSecret      *corev1.Secret
+		imagePullSecrets []corev1.LocalObjectReference
+		wantErr          bool
+	}{{
+		name:             "correct secret",
+		basicSecret:      generateSecret(u.Host, username, password),
+		imagePullSecrets: []corev1.LocalObjectReference{{Name: "secret"}},
+		wantErr:          false,
+	}, {
+		name:             "unauthorized secret",
+		basicSecret:      generateSecret(u.Host, username, "wrong password"),
+		imagePullSecrets: []corev1.LocalObjectReference{{Name: "secret"}},
+		wantErr:          true,
+	}, {
+		name:             "empty secret",
+		basicSecret:      generateSecret(u.Host, "", ""),
+		imagePullSecrets: []corev1.LocalObjectReference{{Name: "secret"}},
+		wantErr:          true,
+	}, {
+		name:             "no basic secret",
+		basicSecret:      &corev1.Secret{},
+		imagePullSecrets: []corev1.LocalObjectReference{{Name: "secret"}},
+		wantErr:          true,
+	}, {
+		name:             "no imagePullSecrets",
+		basicSecret:      generateSecret(u.Host, username, password),
+		imagePullSecrets: []corev1.LocalObjectReference{},
+		wantErr:          true,
+	}} {
```

Comment:
Another useful test I thought of - `[]corev1.LocalObjectReference{badsecret, goodsecret}`

e.g. test that we iterate through the chain properly.

Reply:
Are we expecting err or not for this case? I tried with this case, but `k8schain.New()` will return err `error creating k8schain: secrets "wrong secret" not found`. This is a little bit anti my intuition.

Reply:
This is good to know (also a good reason to add tests for these cases)!

I think it's reasonable to error if the secret doesn't exist. But if the bad secret exists but just isn't the right password, then I would assume the next secret would be tried (otherwise there isn't really a reason to provide multiple secrets).

Reply:
ah I see. Thanks!

Reply:
- So I tested this and found out that {badsecret, goodsecret} cannot work but {goodsecret,badsecret} works. The direct reason is that only 1 auth is written into the request header and sent to the server. The others are not considered. 

After taking a look at the [go-containerregistry](https://github.com/google/go-containerregistry/blob/main/pkg/authn/kubernetes/keychain.go) latest code, I found out that the function use a map to store all the auths, where `k` is the remote host and value is encoded auth. So for this test case. The 2 secrets have the same auth, the first come auth will be stored in keychain and others will be ignored.
  
Related code see here:
https://github.com/google/go-containerregistry/blob/9c35968ef47ec2e4a14a82ea72167f8655c7c853/pkg/authn/kubernetes/keychain.go#L106-L133

- I think this probably makes sense, meanwhile I'm wondering if this test case is valid or not. The secret is created from .`docker/config.json`, the file is written only when user has successfully login. So it is of low possibility that the secret is bad or incorrect. (e.g. user manually manipulates the file later.) 

What do you think?

Reply:
Got it - makes sense once you know the background. I'm neutral if you want to keep the test in or remove it, but we should make sure to document this since it's non-obvious behavior.

Reply:
Sure! I perfer to keep it and document it in the comments.

Reply:
emmm, it may need more restructuring of the code, I will remove it for now

Reply:
> The secret is created from .docker/config.json, the file is written only when user has successfully login.

FWIW, this isn't technically true. You can edit that file to contain any username/password, and many tools implement "login" simply be editing that file without validating that the username/password is correct at all.

If there are improvements we can make to that keychain, especially if it helps Tekton users, we should consider them. I'll happily approve :)

Reply:
> FWIW, this isn't technically true. You can edit that file to contain any username/password, and many tools implement "login" simply be editing that file without validating that the username/password is correct at all.

Oh I see, thank you! I thought the login will do authentication to make sure the auth is valid.

> If there are improvements we can make to that keychain, especially if it helps Tekton users, we should consider them. I'll happily approve 

So this is an interesting case and I'm not sure which makes more sense (Sorry again I'm still new to this but I'm willing to learn more 😄 ). 
From the following code, Say if I have 2 secrets with the same remote registry, but one is incorrect and another is correct. If the bad secret comes first then the correct one will be ignored and not stored in map. Only bad secret will be written into header and sent to server. 

This is just what I see from the test case, what I cannot see is what is the real case, will this {bad secret, correct secret} happen in real use cases? If so, then I think this should be something done from the `go-containerregistry `.


https://github.com/google/go-containerregistry/blob/9c35968ef47ec2e4a14a82ea72167f8655c7c853/pkg/authn/kubernetes/keychain.go#L106

Reply:
Yeah, in that case, today, go-containerregistry will select the first matching secret and use it, not knowing whether it's bad or good. It will only be able to tell whether one is "bad" or "good" by using it to try to auth to the registry, and by that time it's already dropped the old possible alternatives, and will just fail with the "bad" credentials.

This is a bit of a known gap in this keychain logic, and we have some ideas about how to fix it, but it's not really something we see come up much in practice today, so it hasn't been a top priority.